File: src/main/java/graphql/nadel/ServiceExecutionResult.java
package graphql.nadel;

import graphql.PublicApi;

import java.util.List;
import java.util.Map;

import static java.util.Collections.emptyList;
import static java.util.Collections.emptyMap;

@PublicApi
public class ServiceExecutionResult {
    private final List<Map<String, Object>> errors;
    private final Map<String, Object> data;
    private final Map<String, Object> extensions;

    public ServiceExecutionResult(Map<String, Object> data, List<Map<String, Object>> errors, Map<String, Object> extensions) {
        this.data = data;
        this.errors = errors == null ? emptyList() : errors;
        this.extensions = extensions == null ? emptyMap() : extensions;
    }

    public ServiceExecutionResult(Map<String, Object> data) {
        this.data = data;
        this.errors = emptyList();
        this.extensions = emptyMap();
    }

    public Map<String, Object> getData() {
        return data;
    }

    public List<Map<String, Object>> getErrors() {
        return errors;
    }

    public Map<String, Object> getExtensions() {
        return extensions;
    }
}

File: lib/src/main/java/graphql/nadel/ServiceExecutionResult.kt
package graphql.nadel

import graphql.incremental.DelayedIncrementalPartialResult
import graphql.incremental.IncrementalPayload
import org.reactivestreams.Publisher

sealed class ServiceExecutionResult {
    abstract val data: MutableMap<String, Any?>
    abstract val errors: MutableList<MutableMap<String, Any?>?>
    abstract val extensions: MutableMap<String, Any?>
}

data class NadelIncrementalServiceExecutionResult(
    override val data: MutableMap<String, Any?> = LinkedHashMap(),
    override val errors: MutableList<MutableMap<String, Any?>?> = ArrayList(),
    override val extensions: MutableMap<String, Any?> = LinkedHashMap(),
    val incremental: List<IncrementalPayload>?,
    val incrementalItemPublisher: Publisher<DelayedIncrementalPartialResult>,
    val hasNext: Boolean,
) : ServiceExecutionResult()

data class NadelServiceExecutionResultImpl @JvmOverloads constructor(
    override val data: MutableMap<String, Any?> = LinkedHashMap(),
    override val errors: MutableList<MutableMap<String, Any?>?> = ArrayList(),
    override val extensions: MutableMap<String, Any?> = LinkedHashMap(),
) : ServiceExecutionResult()
